src/expr/proof_skolem_cache.h

```diff
@@ -0,0 +1,123 @@
+ *
+ * It is implemented by mapping terms to an attribute corresponding to their
+ * "witness form" as described below. Hence, this class does not impact the
+ * reference counting of skolem variables which may be deleted if they are not
+ * used.
+ */
+class ProofSkolemCache
+{
+ public:
+  ProofSkolemCache() {}
+  ~ProofSkolemCache() {}
+  /**
+   * This makes a skolem of same type as bound variable v, (say its type is T),
+   * whose definition is (witness ((v T)) pred). This definition is maintained
+   * by this class.
+   *
+   * Notice that (exists ((v T)) pred) should be a valid formula. This fact
+   * captures the reason for why the returned Skolem was introduced.
+   *
+   * Take as an example extensionality in arrays:
+   *
+   * (declare-fun a () (Array Int Int))
+   * (declare-fun b () (Array Int Int))
+   * (assert (not (= a b)))
+   *
+   * To witness the index where the arrays a and b are disequal, it is intended
+   * we call this method on:
+   *   Node k = mkSkolem( x, F )
+   * where F is:
+   *   (=> (not (= a b)) (not (= (select a x) (select b x))))
+   * and x is a fresh bound variable of integer type. Internally, this will map
+   * k to the term:
+   *   (witness ((x Int)) (=> (not (= a b))
+   *                          (not (= (select a x) (select b x)))))
+   * A lemma generated by the array solver for extensionality may safely use
+   * the skolem k in the standard way:
+   *   (=> (not (= a b)) (not (= (select a k) (select b k))))
+   * Furthermore, notice that the following lemma does not involve fresh
+   * skolem variables and is valid according to the theory of arrays extended
+   * with support for witness:
+   *   (let ((w (witness ((x Int)) (=> (not (= a b))
+   *                                   (not (= (select a x) (select b x)))))))
+   *     (=> (not (= a b)) (not (= (select a w) (select b w)))))
+   * This version of the lemma, which requires no explicit tracking of free
+   * Skolem variables, can be obtained by a call to getWitnessForm(...)
+   * below. We call this the "witness form" of the lemma above.
+   *
+   * @param v The bound variable of the same type of the Skolem to create.
+   * @param pred The desired property of the Skolem to create, in terms of bound
+   * variable v.
+   * @param prefix The prefix of the name of the Skolem
+   * @param comment Debug information about the Skolem
+   * @param flags The flags for the Skolem (see NodeManager::mkSkolem)
+   * @return The skolem whose witness form is registered by this class.
+   */
+  static Node mkSkolem(Node v,
+                       Node pred,
+                       const std::string& prefix,
+                       const std::string& comment = "",
+                       int flags = NodeManager::SKOLEM_DEFAULT);
+  /**
+   * Same as above, but for special case for (witness ((x T)) (= x t))
+   * where T is the type of t. This skolem is unique for each t.
+   */
+  static Node mkPurifySkolem(Node t,
```

Comment:
I think it'd be good to say here why there is a dedicated attribute for this case.

Reply:
You didn't address this comment.